src/hugr/views/sibling.rs

```diff
@@ -367,6 +364,125 @@ impl<'g, Base: HugrInternals> SiblingSubgraph<'g, Base> {
     }
 }
 
+/// Whether a subgraph is valid.
+///
+/// Does NOT check for convexity.
+fn validate_subgraph<H: HugrView>(
+    hugr: &H,
+    nodes: &[Node],
+    inputs: &IncomingPorts,
+    outputs: &OutgoingPorts,
+) -> Result<(), InvalidSubgraph> {
+    // Check nodes is not empty
+    if nodes.is_empty() {
+        return Err(InvalidSubgraph::EmptySubgraph);
+    }
+    // Check all nodes share parent
+    if !nodes.iter().map(|&n| hugr.get_parent(n)).all_equal() {
+        return Err(InvalidSubgraph::NoSharedParent);
+    }
+
+    // Check inputs are incoming ports and outputs are outgoing ports
+    if inputs
+        .iter()
+        .flatten()
+        .any(|(_, p)| p.direction() == Direction::Outgoing)
+    {
+        return Err(InvalidSubgraph::InvalidBoundary);
+    }
+    if outputs
+        .iter()
+        .any(|(_, p)| p.direction() == Direction::Incoming)
+    {
+        return Err(InvalidSubgraph::InvalidBoundary);
+    }
+
+    let mut ports_inside = inputs.iter().flatten().chain(outputs).copied();
+    let mut ports_outside = ports_inside
+        .clone()
+        .flat_map(|(n, p)| hugr.linked_ports(n, p));
+    // Check incoming & outgoing ports have target resp. source inside
+    if ports_inside.any(|(n, _)| !nodes.contains(&n)) {
+        return Err(InvalidSubgraph::InvalidBoundary);
+    }
+    // Check incoming & outgoing ports have source resp. target outside
+    if ports_outside.any(|(n, _)| nodes.contains(&n)) {
+        return Err(InvalidSubgraph::InvalidBoundary);
+    }
+
+    // Check inputs are unique
+    if !inputs.iter().flatten().all_unique() {
+        return Err(InvalidSubgraph::InvalidBoundary);
+    }
+
+    // Check no incoming partition is empty
+    if inputs.iter().any(|p| p.is_empty()) {
+        return Err(InvalidSubgraph::InvalidBoundary);
+    }
+
+    // Check edge types are equal within partition
```

Comment:
Should we check if the type is copyable here?

Reply:
Done!